src/app/frontend/deploy/publishfromsettings_controller.js

```diff
@@ -0,0 +1,119 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {stateName as replicasetliststate} from 'replicasetlist/replicasetlist_state';
+
+/**
+ * Controller for the publish from settings directive.
+ *
+ * @final
+ */
+export default class PublishFromSettingsController {
+  /**
+   * @param {!angular.$log} $log
+   * @param {!ui.router.$state} $state
+   * @param {!angular.$resource} $resource
+   * @param {!angular.$q} $q
+   * @ngInject
+   */
+  constructor($log, $state, $resource, $q) {
+    /** @export {Object} */
+    this.publisher = this;
+
+    /** @private {!angular.$q} */
+    this.q_ = $q;
+
+    /** @private {!angular.$resource} */
+    this.resource_ = $resource;
+
+    /** @private {!angular.$log} */
+    this.log_ = $log;
+
+    /** @private {!ui.router.$state} */
+    this.state_ = $state;
+
+    /** @export {string} */
+    this.containerImage = '';
+
+    /** @export {number} */
+    this.replicas = 1;
+
+    /**
+     * List of supported protocols.
+     * TODO(bryk): Do not hardcode the here, move to backend.
+     * @const @export {!Array<string>}
+     */
+    this.protocols = ['TCP', 'UDP'];
+
+    /** @export {!Array<!backendApi.PortMapping>} */
+    this.portMappings = [this.newEmptyPortMapping_(this.protocols[0])];
+
+    /** @export {boolean} */
+    this.isExternal = false;
+  }
+
+  /**
+   * Deploys the application based on the state of the controller.
+   *
+   * @export
+   * @return {angular.$q.Promise}
+   */
+  deploy() {
+    let d = this.q_.defer();
```

Comment:
Can you move this variable as far to the bottom of this method as possible? It is a good programming practice to have var initialization as close to its use as possible.

Also, re name: how about naming it deferred? We don't need to save a few keystrokes and `d` means little